core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala

```diff
@@ -369,6 +370,32 @@ class GroupCoordinator(val brokerId: Int,
     }
   }
 
+  /**
+   * try to complete produce, fetch and delete requests if the HW of partition is incremented. Otherwise, we try to complete
+   * only delayed fetch requests.
+   *
+   * Noted that this method may hold a lot of group lock so the caller should NOT hold any group lock
+   * in order to avoid deadlock
+   * @param topicPartitions a map contains the partition and a flag indicting whether the HWM has been changed
+   */
+  private[group] def completeDelayedRequests(topicPartitions: Map[TopicPartition, LeaderHwChange]): Unit =
+    topicPartitions.foreach {
+      case (tp, leaderHWIncremented) => leaderHWIncremented match {
+        case LeaderHwIncremented$ => groupManager.replicaManager.completeDelayedRequests(tp)
```

Comment:
I notice that we are including the `$` here and in a few other places, we should not do that.

Reply:
will copy that!